src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,79 @@ def self.compile(compile_options)
       end
 
       def self.compile_ast(code, compile_options)
+        # TODO: move requires to optimal place and guard against FOSS not having bolt lib code
+        # CODEREVIEW: where should these requires be, and what should happen if they are not found?
+        require 'bolt'
+        require 'bolt/target'
+        require 'bolt/apply_inventory'
+        require 'bolt/apply_target'
+        require 'bolt/apply_result'
+        require 'addressable'
+
+        Puppet[:rich_data] = true
         Puppet[:node_name_value] = compile_options['certname']
 
+        # TODO: again, error handling/warning here when this is not supplied
+        boltlib_path = Array(compile_options.dig('options', 'boltlib') || '')
         # Use the existing environment with the requested name
         Puppet::Pal.in_environment(compile_options['environment'],
+                                   pre_modulepath: boltlib_path,
                                    envpath: Puppet[:environmentpath],
-                                   facts: compile_options['facts']['values'],
-                                   variables: compile_options['variables']['values']) do |pal|
```

Comment:
What was being done with these `variables` before? Is anything expecting them?

Reply:
So i think this is fundamentally something that @justinstoller was asking about later in https://github.com/puppetlabs/puppetserver/pull/2329/files#r416030694 I think the `Puppet::Pal#in_environment` method was designed for the caller to have already defined a full set of variables up front. I *think* we need to be in a compiler for https://github.com/puppetlabs/puppetserver/pull/2329/files#diff-73104a90aa2b4d7f02119c3f544b30c6R67 to do what we want. And we need to resolve those vars to override ones that would otherwise conflict with facts. I can experiment with resolving the conflicts before the call to in_environment.

Reply:
Yeah, it looks like @lucywyman tried to get away with it outside the catalog compiler and it didnt work. https://github.com/puppetlabs/bolt/blame/master/lib/bolt/catalog.rb#L73-L74